### PR TITLE
Fix MSVC build with CUDA enabled

### LINKED ISSUE #115922: Pytorch build from source fails
If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description
Pytorch build from source fails
Provide a short description.

## Code example

``
Please try to provide a minimal example to repro the bug.
Error messages and stack traces are also helpful.

## System Info
Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:

wget https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
base) E:\ons_5\ons_py_torch\ons_pytorch_build>python setup.py develop
Building wheel torch-2.3.0a0+gitUnknown
-- Building version 2.3.0a0+gitUnknown
cmake --build . --target install --config Release
[9/5776] Linking CXX static library lib\libprotobuf-lite.lib
FAILED: lib/libprotobuf-lite.lib
cmd.exe /C "cd . && E:\PROGRA~2\MICROS~3\2022\BUILDT~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x64\lib.exe /nologo /machine:x64 /ignore:4049 /ignore:4217 /ignore:4099 /ignore:4221 /out:lib\libprotobuf-lite.lib third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\any_lite.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\arena.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\extension_set.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\generated_enum_util.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\generated_message_table_driven_lite.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\generated_message_util.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\implicit_weak_message.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\io\coded_stream.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\io\io_win32.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\io\strtod.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\io\zero_copy_stream.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\io\zero_copy_stream_impl.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\io\zero_copy_stream_impl_lite.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\message_lite.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\parse_context.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\repeated_field.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\bytestream.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\common.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\int128.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\status.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\statusor.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\stringpiece.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\stringprintf.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\structurally_valid.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\strutil.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\stubs\time.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\__\src\google\protobuf\wire_format_lite.cc.obj third_party\protobuf\cmake\CMakeFiles\libprotobuf-lite.dir\version.rc.res  && C:\ProgramData\Anaconda3\Library\bin\cmake.exe -E __run_co_compile --lwyu=lib\libprotobuf-lite.lib && cd ."
Error running 'ldd': The system cannot find the file specified
[14/5776] Building CXX object third_party\protobuf\cmake\C...le\protobuf\compiler\csharp\csharp_reflection_class.cc.obj
ninja: build stopped: subcommand failed.

- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): source
- Build command you used (if compiling from source):  python setup.py develop
- OS: win 10
- PyTorch version:  torch version 2.3.0
- Python version: 3.9.7
- CUDA/cuDNN version: not used
- GPU models and configuration: not used
- GCC version (if compiling from source): visual studio build tools
- CMake version:  3.9.6
- Versions of any other relevant libraries:
Kind regards,
Manu

cc @malfet @seemethere @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite

### PR DESCRIPTION
**Pull request by pbelcak**
Fixes #115922 and #101135 

cc @ptrblck @msaroufim @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by pbelcak** [comment]
Also fixes #116242

**Comment by Skylion007** [comment]
@pbelcak For the NVTX issue, you should probably take a look at this PR: https://github.com/pytorch/pytorch/pull/109843 . It tries to add NVTK3 as a PyTorch submodule, but was never merged. cc: @cyyever

**Comment by Skylion007** [comment]
@atalman Is the signal for NVTX3 Green Internally yet? https://github.com/pytorch/pytorch/pull/109843

**Comment by pbelcak** [comment]
> @atalman Is the signal for NVTX3 Green Internally yet? #109843

Not sure if this PR depends on it (though it might); building pytorch with the CUDA Toolkit 12.x installed out of the box simply fails on Windows because the cudaToolsExt libraries are simply no longer there; the PR makes changes that alleviate this.

File: aten/src/ATen/cpu/vec/vec_base.h
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec_base.h:150]: Do you mind deleting this line altogether? And half of the preamble about not supporting C++17 yet?

**Review by malfet — CHANGES_REQUESTED** [review]
Thank you for the change, but do you mind splitting this one int two PRs:
 - One that fixes compilation problem in vec_base.h (please mention the compiler error that were generated before it applied)
 - Another that fixes nvtx discovery (though I it would be better to modify torch:nvtoolsext discovery logic with trying to find it in the cuda toolkit folder and then fallback to the one embedded into the nccl, rather than add `CUDA_VERSION_MAJOR > 12` condition, which is not actually correct, nvtx is shipped with the toolkit but no longer installed by default)

**Comment by cyyever** [comment]
I can't find cudaToolsExt on Windows too, that is why I was pushing the PR. Unfortunately, some META builds should be fixed before it can be landed.

**Comment by patrikhuber** [comment]
> Thank you for the change, but do you mind splitting this one int two PRs:
> 
> * One that fixes compilation problem in vec_base.h (please mention the compiler error that were generated before it applied)
> * Another that fixes nvtx discovery (though I it would be better to modify torch:nvtoolsext discovery logic with trying to find it in the cuda toolkit folder and then fallback to the one embedded into the nccl, rather than add `CUDA_VERSION_MAJOR > 12` condition, which is not actually correct, nvtx is shipped with the toolkit but no longer installed by default)

Hi,

I've been encountering both of these issues too and I can provide a bit more information on them.

1) The latest MSVC produced an ICE (internal compiler error). It looks like the workaround in vec_base that was originally needed for an earlier version of MSVC is now no longer needed in the latest MSVC, and in fact creates an ICE. I've logged this with the Visual Studio team https://developercommunity.visualstudio.com/t/post/10539393 but the easiest is, like done in this PR, just to remove the workaround.

2) NVTX **are** indeed still present in the latest CUDA versions, but they're no longer a compiled library: It's now a header-only library. That's why there isn't a .lib file anymore. See here for more details: https://nvidia.github.io/NVTX/. Thus, I believe the best way forward is probably to update libtorch's CMake scripts so that they now look for the header-only version of NVTX if libtorch needs it. Or obviously if NVTX is not needed to build libtorch, it should be removed as a dependency in libtorch's CMake scripts.

**Comment by patrikhuber** [comment]
And just in case it's helpful, it's not a proper fix I believe, but I made the NvToolsExt issue go away by:

Adding in cuda.cmake L59 after find_package(CUDAToolkit REQUIRED):
```
add_library(CUDA::nvToolsExt INTERFACE IMPORTED)
set_property(TARGET CUDA::nvToolsExt APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${CUDAToolkit_nvToolsExt_INCLUDE_DIRS}")
```

And in share/cmake/Torch/TorchConfig.cmake, this needs to be commented in the if(MSVC)... branch:
```
    set(TORCH_CUDA_LIBRARIES
      # ${NVTOOLEXT_HOME}/lib/x64/nvToolsExt64_1.lib
      ${CUDA_LIBRARIES})
```

And then I think I also needed to add this environment variable in Windows:
`NVTOOLSEXT_PATH = C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.1\`

A dirty workaround for sure but perhaps it helps someone to craft the right PR.

**Comment by patrikhuber** [comment]
Any update on the NVTX issue?

See my comment here for what's happening:

> NVTX are indeed still present in the latest CUDA versions, but they're no longer a compiled library: It's now a header-only library. That's why there isn't a .lib file anymore. See here for more details: https://nvidia.github.io/NVTX/. Thus, I believe the best way forward is probably to update libtorch's CMake scripts so that they now look for the header-only version of NVTX if libtorch needs it. Or obviously if NVTX is not needed to build libtorch, it should be removed as a dependency in libtorch's CMake scripts.

**Comment by cyyever** [comment]
@patrikhuber We had some troubles when imported the previous NVTX PR into META since some inner builds failed.

**Comment by patrikhuber** [comment]
> @patrikhuber We had some troubles when imported the previous NVTX PR into META since some inner builds failed.

I'm not sure what you mean with "previous NVTX PR" but I'm not necessarily advocating to merge this PR, I'm just saying this is quite a big issue (basically it breaks using/importing the library with CUDA 12 on Windows) and it would be nice to get it fixed.

**Comment by cyyever** [comment]
@patrikhuber  I meant #109843

**Comment by ricardoassaad** [comment]
> And just in case it's helpful, it's not a proper fix I believe, but I made the NvToolsExt issue go away by:
> 
> Adding in cuda.cmake L59 after find_package(CUDAToolkit REQUIRED):
> 
> ```
> add_library(CUDA::nvToolsExt INTERFACE IMPORTED)
> set_property(TARGET CUDA::nvToolsExt APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${CUDAToolkit_nvToolsExt_INCLUDE_DIRS}")
> ```
> 
> And in share/cmake/Torch/TorchConfig.cmake, this needs to be commented in the if(MSVC)... branch:
> 
> ```
>     set(TORCH_CUDA_LIBRARIES
>       # ${NVTOOLEXT_HOME}/lib/x64/nvToolsExt64_1.lib
>       ${CUDA_LIBRARIES})
> ```
> 
> And then I think I also needed to add this environment variable in Windows: `NVTOOLSEXT_PATH = C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.1\`
> 
> A dirty workaround for sure but perhaps it helps someone to craft the right PR.

in Cuda 11.8 nvtoolsext is also a header file. the lib file does not exist. I tried your workaround but now i get this error:

Severity	Code	Description	Project	File	Line	Suppression State
Error		CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Torch (missing: TORCH_LIBRARY)		C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake	230

**Comment by patrikhuber** [comment]
The proper way forward with this is that PyTorch has to use `find_package(CUDAToolkit)` and its `CUDA::nvtx3` target - see here https://gitlab.kitware.com/cmake/cmake/-/issues/22893#note_1503430. @malfet @atalman who is the PyTorch CMake maintainer?

**Comment by ricardoassaad** [comment]
> And just in case it's helpful, it's not a proper fix I believe, but I made the NvToolsExt issue go away by:
> 
> Adding in cuda.cmake L59 after find_package(CUDAToolkit REQUIRED):
> 
> ```
> add_library(CUDA::nvToolsExt INTERFACE IMPORTED)
> set_property(TARGET CUDA::nvToolsExt APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${CUDAToolkit_nvToolsExt_INCLUDE_DIRS}")
> ```
> 
> And in share/cmake/Torch/TorchConfig.cmake, this needs to be commented in the if(MSVC)... branch:
> 
> ```
>     set(TORCH_CUDA_LIBRARIES
>       # ${NVTOOLEXT_HOME}/lib/x64/nvToolsExt64_1.lib
>       ${CUDA_LIBRARIES})
> ```
> 
> And then I think I also needed to add this environment variable in Windows: `NVTOOLSEXT_PATH = C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.1\`
> 
> A dirty workaround for sure but perhaps it helps someone to craft the right PR.

this worked for me. i am on cuda 11.8. I only needed to change the NVTOOLEXT_HOME and change this line list(APPEND TORCH_INCLUDE_DIRS ${NVTOOLEXT_HOME}) under the MSVC section in the TorchConfig.cmake. Thanks

**Comment by cyyever** [comment]
@patrikhuber That means we need to jump CMake version, which may affect packaging.

**Comment by patrikhuber** [comment]
> @patrikhuber That means we need to jump CMake version, which may affect packaging.

It looks like the nvtx3 target was added in 3.25 https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#findcudatoolkit.
In any case, if you don't want to bump the CMake version, some patch is needed to fix the problem.

**Comment by patrikhuber** [comment]
I guess if https://github.com/pytorch/pytorch/pull/109843 could be merged, then this could be closed. The other MSVC build problem that this PR tackled was already fixed and merged here https://github.com/pytorch/pytorch/pull/117497.

@pbelcak maybe you want to have a look at https://github.com/pytorch/pytorch/pull/109843?

In any case, @malfet, it would be really great if the NVTX CMake issues could finally be solved, merged, and released.

**Comment by aleksandaratanasov** [comment]
Bumping up the CMake version doesn't help since the affected `TorchConfig.cmake`, Caffe2's `cuda.cmake` and perhaps other files depending on the `nvToolsExt` binary library are shipped with the download of `libtorch` 11.8, 12.1 and possibly other versions.

In `cuda.cmake` from Caffe2 (found inside `share\cmake\Caffe2\public` of the unpacked libtorch ZIP archive) the dependency is on `CUDA::nvToolsExt`.

**Comment by aleksandaratanasov** [comment]
I have an older download of `libtorch 2.1.0` where the DLL `nvToolsExt64_1.dll` can be found inside `lib` of the root directory of `libtorch` (wherever you have extracted the ZIP/tar archive).

**Comment by aleksandaratanasov** [comment]
[This](https://discuss.pytorch.org/t/failed-to-find-nvtoolsext/179635/4) points a working solution (at least in my case) that does not involve copying DLLs from somewhere. The installer of CUDA 11.8 allows to select only the NXVT component that is then installed at `C:/Program Files/NVIDIA Corporation/NvToolsExt`.

**Pull request closed without merging**
